OData/src/System.Web.OData/OData/Routing/Conventions/ProcedureRoutingConventionHelpers.cs

```diff
@@ -54,15 +56,94 @@ public static string SelectAction(this IEdmOperation operation, ILookup<string,
             return null;
         }
 
-        public static void AddKeyValueToRouteData(this HttpControllerContext controllerContext, ODataPath odataPath)
+        public static void AddKeyValueToRouteData(this HttpControllerContext controllerContext, KeyValuePathSegment keySegment, IEdmEntityType entityType, string keyPrefix)
         {
             Contract.Assert(controllerContext != null);
-            Contract.Assert(odataPath != null);
+            Contract.Assert(keySegment != null);
 
-            KeyValuePathSegment keyValueSegment = odataPath.Segments[1] as KeyValuePathSegment;
-            if (keyValueSegment != null)
+            IDictionary<string, object> routingConventionsStore = controllerContext.Request.ODataProperties().RoutingConventionsStore;
+
+            if (entityType == null && keySegment.Segment != null)
             {
-                controllerContext.RouteData.Values[ODataRouteConstants.Key] = keyValueSegment.Value;
+                entityType = keySegment.Segment.EdmType as IEdmEntityType;
+            }
```

Comment:
added

---
In reply to: [63321994](https://github.com/OData/WebApi/pull/727#discussion_r63321994) [](ancestors = 63321994)